examples/rail_car_agent.py

```diff
@@ -0,0 +1,88 @@
+# BSD 3-Clause License
+#
+# Copyright (c) 2023, Woven by Toyota
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import os
+
+from maliput_sim.core.sim import SimulationConfig, AgentInitialState
+from maliput_sim.simulation import Simulation
+
+import maliput.plugin
+import maliput.utility
+
+from controllers import RailCarController
+
+def main():
+    """Main entry point."""
+
+    # Create road network model
+    env = os.getenv("MALIPUT_MALIDRIVE_RESOURCE_ROOT").split(":")
+    env = env[0]
+    print(env)
+    rn_configuration = {"opendrive_file": env +
+                        "/resources/odr/FlatTown01.xodr" if env is not None else "",
+                        "road_geometry_id": "FlatTown01_RoadNetwork"}
+    maliput_road_network = maliput.plugin.create_road_network(
+        "maliput_malidrive", rn_configuration)
+
+    # Sim config
+    sim_config = SimulationConfig(real_time_factor=0., time_step=0.05)
+
+    # Create simulation instance
+    sim = Simulation(maliput_road_network, sim_config)
+
+    # Add agents.
+    # AgentState: name, position, rotation, linear_vel, angular_vel
+
+    initial_agent_state_1 = AgentInitialState(
+        "agent_1", [20., -1., 0.], [0., 0., 0., 1], 10., 0.)
+    rail_car_1 = RailCarController()
+    sim.add_agent(initial_agent_state_1, rail_car_1.step, rail_car_1.get_state)
```

Comment:
Now that I see it separated, why not making something like:

```python
class RailcarController(Behavior):
...
sim.add_agent(initial_agent_state_1, RailcarController())
```

Some comments about this:
- The AgentInitialState is not the same content wise to the output dictionary that the RailcarController has. Thus, I see it more like the controller initialization rather than the state. Nitpick though.
- I think it'll be more flexibile to compose the get_state and step functions within the simulator but to be honest we will require that state relationship between the two functions so I think the cost of implementing a Behavior are minimal w.r.t. to the added complexity of the flexibility. Also, by doing so we get lifecycle management of all the controllers by default.

Reply:
I like this approach and actually this was my first approach, sorry I haven't commented here why I took another path.

The problem of doing this (probably solvable tho) relies on how the simulation stores the state after each step. 
Basically, the ecm will go over all the entities and their components and execute the `get_state` method they have.
So far so good, however, as the `Behavior` component is treated the same way as any other component when the states of the components are obtained: The entities 'components are stored in a dict(list) using the component type(type().__name__) as key and the list of component's states of that type as value. 

This causes that the key for example in this case will be `RailcarController` and even though this is a component I'd like to have this under `Behavior` because having custom components as behaviors dificults the visualization or other down-the-river processing that expects generic naming. (I simply go over Behavior components.)

I'd like though the idea of directly inheriting from Behavior, I should revisit how to achieve this

Reply:
Let' save this comment for future reviews of this.
Fine to move forward as is.